### PR TITLE
OpInfo for `*_like` functions

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #66358 OpInfos for: normal, bernoulli, multinomial
* **#65941 OpInfo for `*_like` functions**
* #64282 OpInfos for some Tensor dtype conversion methods

OpInfos for: empty_like, zeros_like, ones_like, full_like, randn_like

Test Plan:
- run tests

Differential Revision: [D31452625](https://our.internmc.facebook.com/intern/diff/D31452625)

### CODE REVIEW

**Comment by zou3519** [comment]
NB: I will fix the lint whenever fb comes back up...

File: torch/testing/_internal/common_methods_invocations.py [hunk 1]
```diff
@@ -2015,6 +2015,27 @@ def sample_inputs_logsumexp(self, device, dtype, requires_grad):
 
     return tuple(samples)
 
+def sample_inputs_like_fns(self, device, dtype, requires_grad, full_like=False):
```
Comment (pmeier) [inline torch/testing/_internal/common_methods_invocations.py:2018]: All sample inputs functions should take `**kwargs`

```suggestion
def sample_inputs_like_fns(self, device, dtype, requires_grad, full_like=False, **kwargs):
```

File: torch/testing/_internal/common_methods_invocations.py [hunk 2]
```diff
@@ -9327,6 +9348,57 @@ def generate_std_var_kwargs(t: torch.Tensor, **kwargs):
...
                DecorateInfo(unittest.skip("Skipped!"), 'TestJit', 'test_variant_consistency_jit'),
            )),
+    OpInfo('empty_like',
+           dtypes=all_types_and_complex_and(torch.bool, torch.half, torch.bfloat16),
+           supports_out=False,
+           sample_inputs_func=sample_inputs_like_fns,
+           supports_autograd=False,
+           skips=(
+               # This method not supported by TorchScript
+               DecorateInfo(unittest.skip("Skipped!"), 'TestJit', 'test_variant_consistency_jit'),
```
Comment (pmeier) [inline torch/testing/_internal/common_methods_invocations.py:9358]: See https://github.com/pytorch/pytorch/pull/64282#discussion_r722917935.

```suggestion
               DecorateInfo(unittest.skip("Skipped!"), 'TestJit', 'test_variant_consistency_jit', dtypes=(torch.float32, torch.complex64)),
```

File: torch/testing/_internal/common_methods_invocations.py [hunk 3]
```diff
@@ -2015,6 +2015,27 @@ def sample_inputs_logsumexp(self, device, dtype, requires_grad):
...
+    )
+    samples = []
+    for shape, kwargs in inputs:
+        t = make_tensor(shape, device, dtype,
+                        low=None, high=None,
+                        requires_grad=requires_grad)
+        if full_like:
+            samples.append(SampleInput(t, args=(123,), kwargs=kwargs))
+        else:
+            samples.append(SampleInput(t, kwargs=kwargs))
```
Comment (pmeier) [inline torch/testing/_internal/common_methods_invocations.py:2035]: I dislike the special casing here for a single op. I would prefer to have a separate sample inputs function for it instead:

```python
def sample_inputs_full_like(*args, fill_value=123, **kwargs):
    sample_inputs = []
    for sample_input in sample_inputs_like_fns(*args, **kwargs):
        sample_input.args = (fill_value,)
        sample_inputs.append(sample_input)
    return sample_inputs
```

**Review by pmeier — COMMENTED** [review]
Some comments inline. Otherwise LGTM when CI is green.

File: torch/testing/_internal/common_methods_invocations.py [hunk 4]
```diff
@@ -2015,6 +2015,27 @@ def sample_inputs_logsumexp(self, device, dtype, requires_grad):
...
     return tuple(samples)
 
+def sample_inputs_like_fns(self, device, dtype, requires_grad, full_like=False):
+    inputs = (
+        ((), {}),
+        ((S, S), {}),
+        ((0, S, 0), {}),
+        ((S,), {'dtype': torch.double}),
+        ((S,), {'device': 'cpu'}),
+        ((S,), {'dtype': torch.double, 'device': 'cpu'}),
```
[the remark this answers is not in the document]
Comment (zou3519) [reply torch/testing/_internal/common_methods_invocations.py:2025]: I wanted to test some cases where the (dtype, device) being passed to e.g. empty_like is different from the input Tensor's (dtype, device)

[the remark this answers is not in the document]
Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:2034]: It'd be cool to conditionally add CUDA too (the function can check if CUDA is available).

File: torch/testing/_internal/common_methods_invocations.py [hunk 5]
```diff
@@ -2021,6 +2021,35 @@ def sample_inputs_logsumexp(self, device, dtype, requires_grad):
 
     return tuple(samples)
 
+def sample_inputs_like_fns(self, device, dtype, requires_grad, **kwargs):
+    inputs: List[Tuple[Tuple, Dict]] = [
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:2025]: Meta note: typing is lame -- did something complain this wasn't done? If so we should fix that so typing is never required in this file.

File: torch/testing/_internal/common_methods_invocations.py [hunk 6]
```diff
@@ -2021,6 +2021,35 @@ def sample_inputs_logsumexp(self, device, dtype, requires_grad):
...
+    samples = []
+    for shape, kwargs in inputs:
+        t = make_tensor(shape, device, dtype,
+                        low=None, high=None,
+                        requires_grad=requires_grad)
+        samples.append(SampleInput(t, kwargs=kwargs))
+
+    return tuple(samples)
+
+def sample_inputs_full_like(self, *args, **kwargs):
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:2045]: Cool

File: torch/testing/_internal/common_methods_invocations.py (hunk 1 above)
Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py]: Some sample inputs take optional kwargs that tests can opportunistically set:

https://github.com/pytorch/pytorch/blob/6d4d636d66774eb803cffedd267849a8b8a3b536/torch/testing/_internal/common_methods_invocations.py#L4353

**Review by mruberry — APPROVED** [review]
Cool!

**Comment by zou3519** [comment]
@zou3519 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31452625).

**Review by pmeier — APPROVED** [review]
I second https://github.com/pytorch/pytorch/pull/65941#discussion_r723576574, otherwise LGTM!

**Comment by zou3519** [comment]
@zou3519 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31452625).

**Comment by zou3519** [comment]
@zou3519 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31452625).

**Comment by zou3519** [comment]
@zou3519 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31452625).

**Comment by zou3519** [comment]
@zou3519 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31452625).

**Pull request closed without merging**